common/finddialog.cpp

```diff
@@ -46,38 +52,157 @@ void FindDialog::languageChange()
 }
 
-void FindDialog::setWarningVisible(bool value)
+void FindDialog::sClose()
 {
-  return _lbWarning->setVisible(value);
+  bool _matchCase = _cbMatchCase->isChecked();
+  bool _matchWord = _cbMatchWord->isChecked();
+  bool _regex = _cbRegex->isChecked();
+  bool _wrapAround = _cbWrapAround->isChecked();
+
+  sSetWarning();
+  _lbCount->clear();
+  
+  reject();
 }
 
-void FindDialog::sClose()
+void FindDialog::sCountMatches()
 {
-  reject();
+  int currCursorPos = _text->textCursor().position();
+  bool direction = _reverseSearch;
+
+  //do forward search from the top of the doc
+  _reverseSearch = false;  sSetFlags();
+  sMoveCursorTo(0);
+ 
+  if(_cbRegex->isChecked())
+  {
+    QRegExp term = QRegExp(_leSearch->text());
+    while(_text->find(term,flags))
+      _matches.append(_text->textCursor().position());
+  }
+  else
+  {
+    QString term = _leSearch->text();
+    while(_text->find(term,flags))
+      _matches.append(_text->textCursor().position());
+  }
+  
+  sMoveCursorTo(currCursorPos); // put cursor back where it was
+  _reverseSearch = direction;   // return search direction to what it was
+  _searchChanged=false;
+}
+
+void FindDialog::sFindPrev()
+{
+  _reverseSearch = true;
+  sFind();
+  _reverseSearch = false;
 }
 
 void FindDialog::sFind()
 {
-  if (_lbWarning->isVisible())
+  sSetWarning(false);
+  sSetFlags();
+  _leSearch->setStyleSheet( QString( "background-color: white"));
+
+  if(_searchChanged)
+    sCountMatches();
+  
+  if(!_matches.count())
+  {
+    _leSearch->setStyleSheet( QString( "background-color: red"));
+    _lbCount->setText("No matches");
+    return;
+  }
+
+  /* There is a special case in which backwards search skips matches. This is seen in: 
+  https://bugreports.qt.io/browse/QTBUG-48035
+  This seems like a rare case to fall into */
+  
+  if(_cbRegex->isChecked())
   {
-    setWarningVisible(false);
-    sMoveCursorToStart();
+    QRegExp term = QRegExp(_leSearch->text());
+    if(!_text->find(term,flags))
+    {
+      sSetWarning(true);
+      if(_cbWrapAround->isChecked())
+      {
+        if(!_reverseSearch)
+          sMoveCursorTo(0);
+        else
+          sMoveCursorTo(_text->document()->characterCount()-1);
+      }
+    }
+  }
+  else
+  {
+    QString term = _leSearch->text();
+    if(!_text->find(term,flags))
+    {
+      sSetWarning(true);
+      if(_cbWrapAround->isChecked())
+      {
+        if(!_reverseSearch)
+          sMoveCursorTo(0);
+        else
+          sMoveCursorTo(_text->document()->characterCount()-1);
+      }
+    }
   }
   
-  QString term = _leSearch->text();
-  if(!_text->find(term))
-    setWarningVisible(true);    
+  // update counter on dialog label
+  if(_matches.contains(_text->textCursor().position()))
+  {
+    int current = _matches.indexOf(_text->textCursor().position())+1; 
+    QString count = "Match " + QString::number(current) + " of " +  QString::number(_matches.count());
```

Comment:
This is a good place to use Qt's translation facilities:
```suggestion
    QString count = tr("Match %1 of %2").arg(current).arg(_matches.count());
```